scripts/jenkins/Jenkinsfile

```diff
@@ -0,0 +1,567 @@
+    ],
+    stretch_minimal: [
+        id: docker_image_prefix + 'debian-stretch-minimal',
+        context: "./scripts/docker/debian/stretch",
+        file: "./scripts/docker/debian/stretch/Dockerfile.minimal",
+    ],
+    stretch_doc: [
+        id: docker_image_prefix + 'debian-stretch-doc',
+        context: "./scripts/docker/debian/stretch",
+        file: "./scripts/docker/debian/stretch/Dockerfile.doc",
+    ],
+    jessie: [
+        id: docker_image_prefix + 'debian-jessie',
+        context: "./scripts/docker/debian/jessie",
+        file: "./scripts/docker/debian/jessie/Dockerfile",
+    ],
+    xenial: [
+        id: docker_image_prefix + 'ubuntu-xenial',
+        context: "./scripts/docker/ubuntu/xenial",
+        file: "./scripts/docker/ubuntu/xenial/Dockerfile",
+    ]
+]
+
+// Define reusable cmake Flag globals
+CMAKE_FLAGS_BASE = [
+    'SITE': '${STAGE_NAME}',
+    'KDB_DB_SYSTEM': '${HOME}/.config/kdb/system',
+    'KDB_DB_SPEC': '${HOME}/.config/kdb/spec',
+    'CMAKE_INSTALL_PREFIX': '${WORKSPACE}/system',
+    'INSTALL_SYSTEM_FILES': 'OFF',
+    'BUILD_DOCUMENTATION': 'OFF'
+]
+
+CMAKE_FLAGS_BUILD_ALL = [
+    'BINDINGS': 'ALL;EXPERIMENTAL;-DEPRECATED',
+    'PLUGINS': 'ALL;EXPERIMENTAL;-DEPRECATED',
+    'TOOLS': 'ALL'
+]
+
+CMAKE_FLAGS_COVERAGE = ['ENABLE_COVERAGE': 'ON']
+
+CMAKE_FLAGS_JAVA = [
+    'JAVA_AWT_INCLUDE_PATH': '/usr/lib/jvm/java-8-openjdk-amd64/include/'
+]
+
+CMAKE_FLAGS_CLANG = [
+    'CMAKE_C_COMPILER': 'clang',
+    'CMAKE_CXX_COMPILER': 'clang++'
+]
+
+CMAKE_FLAGS_ASAN = ['ENABLE_ASAN': 'ON']
+CMAKE_FLAGS_DEBUG = ['ENABLE_DEBUG': 'ON']
+CMAKE_FLAGS_LOGGER = ['ENABLE_LOGGER': 'ON']
+
+// Define Test Flag globals
+TEST_MEM = 'mem'
+TEST_NOKDB = 'nokdb'
+TEST_ALL = 'all'
+TEST_INSTALL = 'install'
+
+// DATE
+NOW = new Date()
+
+/* Determine what needs to be build
+ *
+ * Execution of this script is triggered by SCM changes or manually by
+ * commenting 'jenkins build Jenkinsfile[REGEX] please. The [REGEX] is
+ * optional but when it exists it allows for specification of which
+ * stages should be build.
+ */
+target = determineBuildTarget()
+
+/* Main Stages
+ *
+ * Serial stages that contain parallized logic. Only proceeds to the next
+ * if previous stage did not fail.
+ */
+lock('docker-images') {
+    stage("Build docker images") {
+        // makes sure docker images corresponding to the Dockerfiles are
+        //   available to run the tests in
+        parallel generate_docker_images()
+    }
+}
+
+stage("Fast builds") {
+    // quick builds to give fast initial feedback to programmers
+    parallel generate_elektra_builds_fast()
+}
+
+stage("Full builds") {
+    // tests to reach full coverage
+    parallel generate_elektra_builds_full()
+}
+
+
+// Stage Generators
+def generate_docker_images() {
+    def tasks = [:]
+
+    DOCKER_IMAGES.each { id, image ->
+        tasks << build_image(image)
+    }
+    return tasks
+}
+
+def build_image(image) {
```

Comment:
which kind of image? why do we need to build it, if we already pass it as parameter?

Reply:
Just because there is an image definition does not mean it has been built yet hence the need to try and build it.

Reply:
The questions should be answered in documentation, please :smile: